src/union.rs

```diff
@@ -10,106 +10,167 @@
 // License for the specific language governing permissions and limitations
 // under the License.
 
-use crate::{digraph, digraph::PyDiGraph};
-use hashbrown::{HashMap, HashSet};
-use petgraph::algo;
-use petgraph::graph::EdgeIndex;
+use crate::{digraph, graph};
+
+use petgraph::stable_graph::{NodeIndex, StableGraph};
+use petgraph::visit::{EdgeRef, IntoEdgeReferences, NodeIndexable};
+use petgraph::{algo, EdgeType};
+
 use pyo3::prelude::*;
 use pyo3::Python;
-use std::cmp::Ordering;
 
-/// [Graph] Return a new PyDiGraph by forming a union from`a` and `b` graphs.
-///
-/// The algorithm has three phases:
-///  - adds all nodes from `b` to `a`. operates in O(n), n being number of nodes in `b`.
-///  - merges nodes from `b` over `a` given that:
-///     - `merge_nodes` is `true`. operates in O(n^2), n being number of nodes in `b`.
-///     - respective node in`b` and `a` share the same weight
-///  - adds all edges from `b` to `a`.
-///     - `merge_edges` is `true`
-///     - respective edge in`b` and `a` share the same weight
-///
-/// with the same weight in graphs `a` and `b` and merged those nodes.
-///
-/// The nodes from graph `b` will replace nodes from `a`.
-///
-///  At this point, only `PyDiGraph` is supported.
-fn _digraph_union(
-    py: Python,
-    a: &PyDiGraph,
-    b: &PyDiGraph,
-    merge_nodes: bool,
-    merge_edges: bool,
-) -> PyResult<PyDiGraph> {
-    let first = &a.graph;
-    let second = &b.graph;
-    let mut combined = PyDiGraph {
-        graph: first.clone(),
-        cycle_state: algo::DfsSpace::default(),
-        check_cycle: false,
-        node_removed: false,
-        multigraph: true,
-    };
-    let mut node_map = HashMap::with_capacity(second.node_count());
-    let mut edge_map = HashSet::with_capacity(second.edge_count());
+type StablePyGraph<Ty> = StableGraph<PyObject, PyObject, Ty>;
 
-    let compare_weights = |a: &PyAny, b: &PyAny| -> PyResult<bool> {
-        let res = a.compare(b)?;
-        Ok(res == Ordering::Equal)
-    };
-
-    for node in second.node_indices() {
-        let node_index = combined.add_node(second[node].clone_ref(py))?;
-        node_map.insert(node.index(), node_index);
+fn find_node_by_weight<Ty: EdgeType>(
```

Comment:
Do we want to reuse this for: https://github.com/Qiskit/retworkx/blob/main/src/digraph.rs#L1400-L1419 (also just noticed it's not handling the python exception correctly)? Also should we expose a similar method for pygraph?

Reply:
Done in 088f27d where i moved this in `lib.rs` so we can reuse it for `PyDiGraph` and `PyGraph` methods.